docs: add documentation on how to install trivy-operator in a ns with default deny-all netpols

francRang opened the pull request:
## Description

## Related issues
- Close #1598 

Remove this section if you don't have related PRs.

## Checklist
- [x] I've read the [guidelines for contributing](https://github.com/aquasecurity/trivy-operator/blob/main/CONTRIBUTING.md) to this repository.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy-operator/tree/main/docs) with the relevant information (if needed).
- [x] I've added usage information (if the PR introduces new options)
- [x] I've included a "before" and "after" example to the description (if the PR is a user interface change).

chen-keinan commented:
@francRang is it ready for review ?

francRang commented:
@chen-keinan 
No, still WIP. Will complete it soon. Hopefully next week once I got access to a computer again.

chen-keinan commented:
@francRang any update on this pr

francRang commented:
> @francRang any update on this pr

I stopped working on it, but can re take it and submit something for review by the end of this week.

francRang commented:
Got most of the scenarios ready, a few more and I shall be done.

francRang commented:
@chen-keinan, PR is ready for review. It addresses the following scenarios:
- Deploying trivy-operator to a default deny-all ingress/egress namespace in standalone mode
- Deploying trivy-operator + trivy (server) in client/server mode
Let me know if there is anything that needs improvement. Feel free to follow the guide to make sure things work as expected.

chen-keinan commented on docs/getting-started/installation/troubleshooting.md:
`````diff
@@ -109,3 +109,338 @@ The operator also could not be configured to scan the workload you are expecting
+
+```
+dial tcp 10.43.158.111:4954: connect: connection refused
+````
`````
there is an extra apostrophe in L417

chen-keinan commented on docs/getting-started/installation/troubleshooting.md:
````diff
@@ -109,3 +109,338 @@ The operator also could not be configured to scan the workload you are expecting
+
+If you are trying to install the Trivy-Operator in an environment where there are default deny-all egress/ingress network policies (see example below), you might need to configure some extra network policies yourself to make sure the traffic can flow as expected and the operator does not enter an error state.
+
+```
````
can you add `yaml` after ``` so it will be formatted as yaml

chen-keinan commented on docs/getting-started/installation/troubleshooting.md:
````diff
@@ -109,3 +109,338 @@ The operator also could not be configured to scan the workload you are expecting
+
+Basically, the trivy-operator cannot reach the kube-api to execute the calls looking for different resources. In example above, we can see that the trivy-operator was looking for the trivy-operator configmap in the trivy-system. Our first task will be to enable traffic between the trivy-operator pods and the kube-api service so that the trivy-operator can successfully get what it needs from the k ...
+
+```
````
can you add `yaml` after ``` so it will be formatted as yaml

chen-keinan commented on docs/getting-started/installation/troubleshooting.md:
````diff
@@ -109,3 +109,338 @@ The operator also could not be configured to scan the workload you are expecting
+
+To remediate this issue, we'll need to create a network policy allowing traffic on port `53` to the `kube-system` namespace. This will allow the trivy-systems to perforn DNS lookups via the `core-dns` pods. 
+
+```
````
can you add `yaml` after ``` so it will be formatted as yaml

chen-keinan commented on docs/getting-started/installation/troubleshooting.md:
````diff
@@ -109,3 +109,338 @@ The operator also could not be configured to scan the workload you are expecting
+
+We proceed to create the network policy as follows:
+
+```
````
can you add `yaml` after ``` so it will be formatted as yaml

chen-keinan commented on docs/getting-started/installation/troubleshooting.md:
````diff
@@ -109,3 +109,338 @@ The operator also could not be configured to scan the workload you are expecting
+
+We can make use of the label `app.kubernetes.io/name=trivy`, so the resulting network policy will look like this:
+
+```
````
can you add `yaml` after ``` so it will be formatted as yaml

chen-keinan commented on docs/getting-started/installation/troubleshooting.md:
````diff
@@ -109,3 +109,338 @@ The operator also could not be configured to scan the workload you are expecting
+
+The trivy-operator has to reach out to the trivy-server on port `4954` in order to access the downloaded vulnerability database. We also need to enable that connection via a networkpolicy (you guessed it), we can delete the previously created network policy `allow-egress-443-trivy-operator` via `kubectl delete networkpolicy allow-egress-443-trivy-operator -n trivy-system` and create a new one wit ...
+
+```
````
can you add `yaml` after ``` so it will be formatted as yaml

chen-keinan commented on docs/getting-started/installation/troubleshooting.md:
````diff
@@ -109,3 +109,338 @@ The operator also could not be configured to scan the workload you are expecting
+
+So far we have created network policies to allow egress traffic. There is one last missing network policy and it is of type ingress. This network policy will allow the trivy-server to receive traffic on port `4954`.
+
+```
````
can you add `yaml` after ``` so it will be formatted as yaml

chen-keinan commented on docs/getting-started/installation/troubleshooting.md:
````diff
@@ -109,3 +109,338 @@ The operator also could not be configured to scan the workload you are expecting
+
+This network policy uses the appropiate `matchLabels` to only target the trivy-server. When restarting the trivy-operator with `kubectl rollout restart deployment trivy-operator -n trivy-system`, we see that the errors are gone. When doing a `kubectl get vulnerabilityreport -n applications`, we see that there is a newly generated `vulnerabilityreport` for our `nginx` pod:
+
+```
````
can you add `sh` after ``` so it will be formatted as shell command

chen-keinan commented on docs/getting-started/installation/troubleshooting.md:
````diff
@@ -109,3 +109,338 @@ The operator also could not be configured to scan the workload you are expecting
+
+At this point, we should expect the Trivy Operator to generate the `VulnerabilityReports` custom resources in our `applications` namespace. However, if we try to get these resources across the `applications` namespace, we'll see that we won't get any reports:
+
+```
````
can you add `sh` after ``` so it will be formatted as shell command

chen-keinan commented on docs/getting-started/installation/troubleshooting.md:
````diff
@@ -109,3 +109,338 @@ The operator also could not be configured to scan the workload you are expecting
+
+If we look at the `get pods` command, the description of the trivy-operator pod and its logs, we'll see some interesting messages that will help us understand why the reports aren't being generated.
+
+```
````
can you add `sh` after ``` so it will be formatted as shell command

chen-keinan commented on docs/getting-started/installation/troubleshooting.md:
````diff
@@ -109,3 +109,338 @@ The operator also could not be configured to scan the workload you are expecting
+trivy-operator-846f8c6446-clzlk   0/1     CrashLoopBackOff   6 (2m41s ago)   8m28s
+```
+
+```
````
can you add `sh` after ``` so it will be formatted as shell command

chen-keinan commented on docs/getting-started/installation/troubleshooting.md:
````diff
@@ -109,3 +109,338 @@ The operator also could not be configured to scan the workload you are expecting
+
+The pod is in a `CrashLoopBackOff` state, and the description confirms that the container is constantly being restarted.
+
+```
````
can you add `sh` after ``` so it will be formatted as shell command

chen-keinan commented on docs/getting-started/installation/troubleshooting.md:
````diff
@@ -109,3 +109,338 @@ The operator also could not be configured to scan the workload you are expecting
+We see that the trivy-operator is correctly configured to watch all namespaces, that means that the `VulnerabilityReports` should be generated across all namespaces, including the `applications` namespace.
+
+The first red flag that something is wrong with the networking configuration can be found in the following message:
+```
````
can you add `sh` after ``` so it will be formatted as shell command

chen-keinan commented on docs/getting-started/installation/troubleshooting.md:
````diff
@@ -109,3 +109,338 @@ The operator also could not be configured to scan the workload you are expecting
+The first red flag that something is wrong with the networking configuration can be found in the following message:
+```
+unable to run trivy operator: failed getting configmap: trivy-operator: Get "https://10.43.0.1:443/api/v1/namespaces/trivy-system/configmaps/trivy-operator": dial tcp 10.43.0.1:443: connect: connection refused
+```
````
can you add `sh` after ``` so it will be formatted as shell command

chen-keinan commented on docs/getting-started/installation/troubleshooting.md:
````diff
@@ -109,3 +109,338 @@ The operator also could not be configured to scan the workload you are expecting
+
+This means that the trivy-operator cannot resolve DNS records. The cause of this is the fact that the traffic to the `kube-dns` service residing in the `kube-system` namespace is disabled because of the `deny-all` egress network policies in the `trivy-system` namespace. We can confirm this by running the following command and to the svc information:
+
+```
````
can you add `sh` after ``` so it will be formatted as shell command

chen-keinan commented on docs/getting-started/installation/troubleshooting.md:
````diff
@@ -109,3 +109,338 @@ The operator also could not be configured to scan the workload you are expecting
+
+We can confirm this by doing a `watch kubectl get pods -n trivy-system` on pods on the trivy-system, then restarting the trivy-operator via `kubectl rollout restart deployment -n trivy-system`:
+
+```
````
can you add `sh` after ``` so it will be formatted as shell command

chen-keinan commented on docs/getting-started/installation/troubleshooting.md:
````diff
@@ -109,3 +109,338 @@ The operator also could not be configured to scan the workload you are expecting
+
+Here we see the scanning pod being spawned. We now must get a good label so we can create a network policy for it. We do so by grabbing the pod name and getting the labels via `yq` while we watch for the pod in another terminal.
+
+```
````
can you add `sh` after ``` so it will be formatted as shell command

chen-keinan commented on docs/getting-started/installation/troubleshooting.md:
````diff
@@ -109,3 +109,338 @@ The operator also could not be configured to scan the workload you are expecting
+
+We use the CIDR `0.0.0.0/0` to denote that we want to allow the target pods to talk to any IP address, and we specify port `443` as the allowed port. If we query for the logs again, we'll see that the error is gone. Moreover, if we do a `kubectl get vulnerabilityreport -n applications`, we'll see that the report for the `nginx` pod has been recently generated:
+
+```
````
can you add `sh` after ``` so it will be formatted as shell command

chen-keinan commented on docs/getting-started/installation/troubleshooting.md:
````diff
@@ -109,3 +109,338 @@ The operator also could not be configured to scan the workload you are expecting
+When deploying the trivy-operator + trivy-server for downloading the vulnerability database, you will need to create similar network policies to the ones created for the trivy-operator as a standalone component.
+After installing trivy-server in the current cluster, the pod entered a status of `CrashLookBackOff`. Upon inspecting the logs, for the trivy-server statefulset, we counter the following error:
+
+```
````
can you add `sh` after ``` so it will be formatted as shell command

chen-keinan commented on docs/getting-started/installation/troubleshooting.md:
````diff
@@ -109,3 +109,338 @@ The operator also could not be configured to scan the workload you are expecting
+
+This means that the trivy-server cannot connect to the image registry over port `443`. This can be fixed by applying a network policy like `allow-egress-443-trivy-operator`, which we created for the trivy-operator, but first we must get the label that will be used to select the pods that the trivy-server generates. We do so by doing a `kubectl get pods trivy-0 -n trivy-system --show-labels`, we o ...
+
+```
````
can you add `sh` after ``` so it will be formatted as shell command

chen-keinan commented on docs/getting-started/installation/troubleshooting.md:
````diff
@@ -109,3 +109,338 @@ The operator also could not be configured to scan the workload you are expecting
+We proceed to restart the trivy-server statefulset with `kubectl rollout restart sts -n trivy-system trivy` and see that the error previously seen is gone. Trivy-server was able to download the DB and is listening on port `4954`.
+
+```
+kubectl logs -n trivy-system statefulset/trivy
````
can you add `sh` after ``` so it will be formatted as shell command

chen-keinan reviewed (commented):
@francRang lgtm 🚀  I have added few nit formatting comments

chen-keinan commented:
Note: you can test the docs by running :  `mage docs:serve` and put `http://localhost:8000/trivy-operator/` in browser

francRang commented:
All concerns have been addressed, fixed some minor typos as well and updated the Title of the section within the doc to reflect its actual purpose. Ran `mage docs:serve` and everything looks good.

I did wanted to mention something, in the future, we could extent this section of the docs by adding more granular netpols. E.g:

`allow-egress-443-trivy-server` allows all traffic to flow through `443`, regardless of the destination. Maybe we could pinpoint some CIDRs to make sure we only allow the desired traffic and nothing else. I was reading the docs for trivy-db, which is the ultimate destination for this netpol, apparently it is hosted on GHCR, those could go in there as the destination along the `kube-api` CIDR. The main reason why I didn't want to add this in there is because some people might be pulling from other registries by mirroring the images, so those CIDRs can change. Either way, something to keep in mind. Maybe future PRs can address this.

chen-keinan reviewed (approved):
@francRang lgtm 🚀  thank you for the contribution
